pagination/paginator.go

```diff
@@ -0,0 +1,118 @@
+package pagination
+
+import (
+	"fmt"
+	"math"
+)
+
+const defaultPageNumber = 1
+const defaultItemsPerPage = 10
+
+// PaginatorInterface defines what makes a paginator
+type PaginatorInterface interface {
+	Init(PageNumber int, ItemsPerPage int, ItemsPerPageLimit int) (err error)
+	Offset() int
+	SetItemsTotal(i int)
+	Validate() (err error)
+}
+
+// Paginator contains all of the basic pagination fields
+type Paginator struct {
+	PageNumber        int
+	PagesTotal        *int
+	ItemsPerPage      int
+	ItemsPerPageLimit int
+	ItemsTotal        *int
+}
+
+// TooManyItemsPerPageError is the error generated when the requested items per page is greater than the max
+type TooManyItemsPerPageError struct {
+	PerPage, MaxPerPage int
+}
+
+func (e *TooManyItemsPerPageError) Error() string {
+	return fmt.Sprintf("The requested number of items per page (%d) is greater than the maximum allowed (%d)", e.PerPage, e.MaxPerPage)
+}
+
+// InvalidPageNumberError is the error generated when the requested page number is invalid
+type InvalidPageNumberError int
+
+func (e InvalidPageNumberError) Error() string {
+	return fmt.Sprintf("The requested page (%d) is not available", int(e))
+}
+
+// InvalidItemsPerPageError is the error generated when the requested items per page is invalid
+type InvalidItemsPerPageError int
+
+func (e InvalidItemsPerPageError) Error() string {
+	return fmt.Sprintf("The requested items per page (%d) is less than 1", int(e))
+}
+
+// New returns a new Paginator whilst calling init
+func New(PageNumber int, ItemsPerPage int, ItemsPerPageLimit int) (p *Paginator, err error) {
```

Comment:
not essential but it seems weird that these are UpperCaseFirst parameters. when this is a local variable that only exists as part of the method, but indicated as being public and exported.